File: windows/net/samples/mux_mp4_file/Options.cs
/*
 *  Copyright (c) mux_mp4_file Primo Software. All Rights Reserved.
 *
 *  Use of this source code is governed by a BSD-style license
 *  that can be found in the LICENSE file in the root of the source
 *  tree.  
*/

using System;
using System.IO;
using System.Collections.Generic;
using CommandLine;
using CommandLine.Text;
using PrimoSoftware.AVBlocks;

namespace MuxMp4FileSample
{
    class Options
    {
        // mux_mp4_file options
        [Option('?', "help", HelpText = "Display this help screen")]
        public bool Help { get; set; }

        [OptionArray('a', "audio", HelpText = "input AAC files. Can be used multiple times")]
        public string[] AudioFiles { get; set; }

        [OptionArray('v', "video", HelpText = "input H264 files. Can be used multiple times")]
        public string[] VideoFiles { get; set; }

        [Option('o', "output", HelpText = "output file")]
        public string OutputFile { get; set; }
        
        // The program parses the command line options and sets these properties
        public bool Error { get; private set; }

        string GetUsage()
        {
            return HelpText.AutoBuild(this, (HelpText current) => HelpText.DefaultParsingErrorsHandler(this, current));
        }

        void PrintUsage()
        {
            Console.WriteLine("Usage: mux_mp4_file --audio <input_AAC> --video <input_AVC> --output <output.mp4>");
            Console.WriteLine(GetUsage());
        }

        void ResetOptions()
        {
            Help = false;
            Error = false;
            AudioFiles = null;
            VideoFiles = null;
            OutputFile = null;
        }

        void SetDefaultOptions()
        {
            string exeDir = Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location);
            string videoFile = Path.Combine(exeDir, "..\\assets\\mov\\big_buck_bunny_trailer_iphone.vid.mp4");
            string audioFile = Path.Combine(exeDir, "..\\assets\\aud\\big_buck_bunny_trailer_iphone.aud.mp4");

            VideoFiles = new string[] { videoFile };
            AudioFiles = new string[] { audioFile };
            OutputFile = Path.Combine(exeDir, "mux_mp4_file.mp4");

            Console.WriteLine("Using default options:\n --audio " + AudioFiles[0] + " --video " + VideoFiles[0] + " --output " + OutputFile);
        }

        public bool Prepare(string[] args)
        {
            ResetOptions();

            if (args.Length < 3)
            {
                SetDefaultOptions();
            }
            else
            {
                if (!CommandLine.Parser.Default.ParseArguments(args, this))
                {
                    Console.WriteLine("Syntax error");
                    PrintUsage();
                    Error = true;
                    return false;
                }
            }

            if (Help)
            {
                PrintUsage();
                return false;
            }

            if (!Validate())
            {
                Console.WriteLine("\nRequired options are not set!");
                PrintUsage();
                Error = true;
                return false;
            }

            return true;
        }

        bool Validate()
        {
            bool res = true;

            Console.WriteLine("Audio files: ");
            if ((AudioFiles != null) && (AudioFiles.Length > 0))
            {
                foreach (string s in AudioFiles)
                {
                    Console.WriteLine("   " + s);
                }
            }
            else
            {
                Console.WriteLine("[not set]");
                res = false;
            }

            Console.WriteLine("Video files: ");
            if ((VideoFiles != null) && (VideoFiles.Length > 0))
            {
                foreach (string s in VideoFiles)
                {
                    Console.WriteLine("   " + s);
                }
            }
            else
            {
                Console.WriteLine("[not set]");
                res = false;
            }

            Console.Write("Output file: ");
            if (OutputFile == null)
            {
                Console.WriteLine("[not set]");
                res = false;
            }
            else
            {
                Console.WriteLine(OutputFile);
            }

            return res;
        }
    }
}

File: windows/cpp/samples/mux_mp4_file/options.cpp
/*
 *  Copyright (c) 2016 Primo Software. All Rights Reserved.
 *
 *  Use of this source code is governed by a BSD-style license
 *  that can be found in the LICENSE file in the root of the source
 *  tree.  
*/

#include "stdafx.h"
#include "options.h"
#include "util.h"
#include <atlstr.h>
#include "program_options.h"

using namespace std;
using namespace primo::avblocks;	
using namespace primo::codecs;	
using namespace primo::program_options;

void help(OptionsConfig<wchar_t>& optionsConfig)
{
    wcout << endl;
    wcout << L"Usage: mux_mp4_file --audio <input_AAC> --video <input_AVC> --output <output.mp4>" << endl;
    doHelp(wcout, optionsConfig);
}

bool validateOptions(Options& opt)
{    
    if (opt.input_audio.empty() && opt.input_video.empty())
    {
        return false;
    }

    if (opt.output_file.empty())
    {
        return false;
    }
    return true;
}

void setDefaultOptions(Options& opt)
{
    opt.input_audio.push_back(getExeDir() + L"\\..\\assets\\aud\\big_buck_bunny_trailer_iphone.aud.mp4");
    opt.input_video.push_back(getExeDir() + L"\\..\\assets\\mov\\big_buck_bunny_trailer_iphone.vid.mp4");
    opt.output_file = getExeDir() + L"\\mux_mp4_file.mp4";
}

ErrorCodes prepareOptions(Options& opt, int argc, wchar_t* argv[])
{
    if (argc < 2)
    {
        setDefaultOptions(opt);
        wcout << L"Using defaults:\n";
        for(std::vector<wstring>::iterator input = opt.input_audio.begin(); input != opt.input_audio.end(); input++)
        {
            wcout << L" --audio " << *input;
        }
        for (std::vector<wstring>::iterator input = opt.input_video.begin(); input != opt.input_video.end(); input++)
        {
            wcout << L" --video " << *input;
        }
        wcout << L" --output " << opt.output_file;
        wcout << endl;
        return Parsed;
    }

    OptionsConfig<wchar_t> optionsConfig;
    optionsConfig.addOptions()
        (L"help,?",			opt.help,								L"")
        (L"audio,a",		opt.input_audio,	vector<wstring>(),	L"input AAC files. Can be used multiple times")
        (L"video,v",        opt.input_video,    vector<wstring>(),  L"input H264 files. Can be used multiple times")
        (L"output,o",		opt.output_file,	wstring(),			L"output file");

    try
    {
        scanArgv(optionsConfig, argc, argv);
    }
    catch (ParseFailure<wchar_t> &ex)
    {
        wcout << ex.message() << endl;
        help(optionsConfig);
        return Error;
    }

    if(opt.help)
    {
        help(optionsConfig);
        return Command;
    }

    if (!validateOptions(opt))
    {
        help(optionsConfig);
        return Error;
    }

    return Parsed;
}
